### ISSUE TITLE
Revisit lifecycle of editor inputs in VSCode

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#99770]
Stumbled upon this code (after chat with @jrieken): 

https://github.com/microsoft/vscode/blob/19dcc7916c1d092329593c7be02829f66d949f9f/src/vs/workbench/contrib/notebook/browser/notebook.contribution.ts#L216-L230

The editor part should already take care of disposal. You can see it here:

https://github.com/microsoft/vscode/blob/19dcc7916c1d092329593c7be02829f66d949f9f/src/vs/workbench/browser/parts/editor/editorGroupView.ts#L535-L537

We call `EditorInput.dispose()` when the last kind of that editor is closed across all groups. This is in the end determined by the `EditorInput.matches()` method and goes like this:
* for all groups
* for all editors in the group
* check if the `editor.matches(editorThatClosed)`

I see that this may result in editor inputs not getting disposed in the following case:
* user opens notebook `foo`
* user splits the editor and a second notebook `foo` is opened
* at this time we have 2 notebook inputs
* user closes one notebook
* we do not dispose the first notebook input because we find that it is still opened (via `EditorInput.matches()`)

This is unfortunate and should be fixed by always ensuring the same editor input is used for the same underlying resource (something we do for files always). The editor part cannot really enforce this, so I wonder if a better idea would be to introduce a method such as `EditorInput#close()` that we always call when an editor is being closed and then each input has to decide how to deal with it.

**Comment by mjbvz** [comment]
@bpasero I believe we avoid this for webviews and custom editors because we create a unique editor input for each copy of the editor that is open (even if they are for the same resource): https://github.com/microsoft/vscode/blob/e015178ac402a929e51860beb6a41a4e4ff8f6da/src/vs/workbench/contrib/customEditor/browser/customEditors.ts#L487

We do this because `WebviewEditorInput` stores the `<webview>` element, which cannot be shared between multiple editors.

**Comment by bpasero** [comment]
@mjbvz so are you implementing the `matches` method in a way that accounts for this? Otherwise your inputs are not getting disposed?

I am really thinking we should have an explicit `close` method on the editor inputs and let inputs deal with the dispose as it best fits. This method could carry an indicator if the input is still opened in any other group to prevent duplicate work.

**Comment by bpasero** [comment]
This is a list of all our editor inputs in VSCode where we should check for adoption:

![image](https://user-images.githubusercontent.com/900690/84403199-b6ce2a80-ac05-11ea-8b59-95980500360f.png)

Maybe @mjbvz you could make a statement about custom editors and webview related ones.

**Comment by jrieken** [comment]
re https://github.com/microsoft/vscode/issues/99770#issuecomment-642719879

So, `EditorInput#close` is called whenever a tab goes away? Is that also true for moving tabs?

**Comment by rebornix** [comment]
@bpasero we can get rid of the listener with this new API 👍 

Speaking of `EditorInput`, we create editor input for every editor tab even if they might be the same resources, which is the same as the CustomEditor/Webview. The major reason for this design is we want to maintain a 1 to 1 mapping between Editor (Tab) to `NotebookEditorWidget`. If there are two editor tabs for the same resource, there will be two `NotebookEditorWidget`. If users move one editor to another group, we will reuse the `NotebookEditorWidget` from previous editor input.

If we are going to share EditorInput between tabs, then we need to know if an editor input will be detached from a tab or not (the case is Move editor from group to another). If I understand correctly, we only need to make little change to how Move Editors work now in`editorGroupView`


https://github.com/microsoft/vscode/blob/master/src/vs/workbench/browser/parts/editor/editorGroupView.ts#L1163-L1181

```diff
import { ENGINE_METHOD_STORE } from "constants";

private doMoveOrCopyEditorAcrossGroups(editor: EditorInput, target: IEditorGroupView, moveOptions: IMoveEditorOptions = Object.create(null), keepCopy?: boolean): void {

    // When moving/copying an editor, try to preserve as much view state as possible
    // by checking for the editor to be a text editor and creating the options accordingly
    // if so
    const options = getActiveTextEditorOptions(this, editor, EditorOptions.create({
        ...moveOptions,
        pinned: true, 										// always pin moved editor
        sticky: !keepCopy && this._group.isSticky(editor)	// preserve sticky state only if editor is moved (https://github.com/microsoft/vscode/issues/99035)
    }));

+    editor.close(this, true);

    // A move to another group is an open first...
    target.openEditor(editor, options, keepCopy ? OpenEditorContext.COPY_EDITOR : OpenEditorContext.MOVE_EDITOR);

    // ...and a close afterwards (unless we copy)
    if (!keepCopy) {
        this.doCloseEditor(editor, false /* do not focus next one behind if any */);
    }
}
```

and in our `NotebookEditorInput`, we can our own mapping between `EditorGroup` and `NotebookEditorWidget`. Moving editor will firstly de-ref the widget, and then in `NotebookEditor.setInput` we can reuse it. Lastly, when an editorInput is disposed, we can then dispose all `NotebookEditorWiget`s it created.

@jrieken does this sound reasonable to you?

**Comment by jrieken** [comment]
> I'll suggest we have a sync online for more detailed discussions.

Yeah - I have ref counted notebook models follow the re-use editor input approach but it's impossible for to get the "share widget" logic to work. The challange is that is all cases `Editor#setInput` is called first, e.g before `group#onWillCloseEditor` or `clearInput` etc. Only the open-editor-overwrite can do that but it has no other way of identifying editors than the input. Which is sometimes shared and sometimes not shared.

**Comment by bpasero** [comment]
@jrieken 

> So, EditorInput#close is called whenever a tab goes away? Is that also true for moving tabs?

Yes, but the logic for moving a tab will always first open it in the second group and then close it in the first group. So you will end up getting a `close` call but `openedInOtherGroups` will be `true` because the tab is already opened there:

![image](https://user-images.githubusercontent.com/900690/84478802-30135f00-ac92-11ea-80f7-46ca05ff0407.png)

It is fine if you decide to NOT share editor inputs, this is not a requirement of the editor model at all. Just files decided to do this. 

I hope the new `close` method makes the lifecycle easier to adopt. The editor part will no longer just dispose any inputs or make assumptions.

**Comment by jrieken** [comment]
> It is fine if you decide to NOT share editor inputs, this is not a requirement of the editor model at all.

Yeah, I came to the same conclusion and I'll drop my current (branched) changes. One challenge is that the workbench default to reuse and that notebooks and custom editor fight against that. So, could we have workbench support here, e.g when splitting an editor check for `EditorInput#clone` and iff that exists call that?

**Comment by bpasero** [comment]
@jrieken 

> One challenge is that the workbench defaults to reuse

Is that only for splitting or are there more places where you see that?

I think it is fine to add more methods to `EditorInput` to solve this, e.g. we already have that method `supportsSplitEditor` and I think we could change it to return the editor input to use when splitting or `undefined` if unsupported 👍

**Comment by jrieken** [comment]
> we already have that method supportsSplitEditor and I think we could change it to return the editor input to use when splitting or undefined if unsupported 👍

Oh, yeah I like that

**Comment by bpasero** [comment]
So after looking into this again going through all editors, I see that this is not so trivial. I still think a `close` method is good and that the editor part should not simply `dispose` editor inputs. And I think it is good to put the `dispose` control into the hands of the editor input implementation.

However, the logic that we have today that we only `dispose` an editor input if it is not opened in other groups is problematic. Specifically this code:

```ts
close(group: GroupIdentifier, openedInOtherGroups: boolean): void {
	if (!openedInOtherGroups) {
		this.dispose();
	}
}
```

**Here is why:**
* today some editors decide to share the same instance across all groups or not, for example files do but notebooks or custom editors don't
* however this is NOT enforced end to end, you can actually open the same editor in another editor group, following these steps [1]
* if an editor decides to NOT share editor inputs across groups, your input will not be disposed when you close it if it is opened in another group (this is essentially https://github.com/microsoft/vscode/issues/99887). you can now in theory fix this by overriding the `close` method, however if you follow steps from [1] you might end up disposing the same editor that is still opened in another group

**[1] Open same editor across groups**
* open an editor such as notebook or custom editor
* open an empty editor group to the side
* focus that empty editor group
* bring up editor history (`Cmd+P`) and select the notebook or custom editor
=> you have now opened the same editor input in another group, even if that editor configured that it does not support splitting

**Bottom line**: a call between custom editor and notebook owners would be good to collect requirements and then go from there. I will hold to push anymore changes on this matter until we decided what is needed.

**Comment by bpasero** [comment]
Actually how would we enforce this, e.g. someone could easily write the following code:

```ts
const input = new MyInput();
editorService.openEditor(input, groupA);
editorService.openEditor(input, groupB);
```

I feel that we need to make sure that editor inputs really remain lightweight and do not carry anything heavy around. They should only be used to instruct the editor part to open something imho.

**Comment by mjbvz** [comment]
Some quick  high level note on expected behavior for webviews and custom editors.

### Webviews

- Every webview is unique.
- There should ever be one instance of a given webview across all editors.
- Splitting is not allowed.
- Once closed, a webview should not be able to be reopened. (This is broken today)


### Custom Editors (that set supportsMultipleEditorsPerDocument: false)

- There should only ever be one custom editor for this resource across all editors.
- Custom Editors can be reopened after they are closed.

### Normal Custom editors (that set supportsMultipleEditorsPerDocument: true)

- There can only be one custom editor for a resource open in a given editor group
- However there can be multiple editors for this resource across all editor groups
- But all of these editors have their own unique state (they share a document model). This is because each editor instance has its own unique `<webview>` element. We try to preserve this webview element as much as possible when the user moves around editors

**Comment by bpasero** [comment]
The editor part changed its strategy for calling `dispose` in a way that should make more sense now:
- previously: `IEditorInput.dispose()` was only called if the input was not opened in another group by calling `IEditorInput.matches()`
- now: `IEditorInput.dispose()` is called if the input is not opened in another group using `===` checks

As such, inputs should now be more aggressively disposed if they decide to not share the same instance across groups.

As such, this special code for custom editors was removed: https://github.com/microsoft/vscode/commit/063ea11611e2194d5496410a5d0544357c7e7a7a
